src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -0,0 +1,477 @@
+                // can turn off filters they don't want and get the right set of items.
+
+                // If we are going to filter everything out, then just preserve the existing
+                // model (and all the previously filtered items), but switch over to soft
+                // selection.
+                selection = hasActiveFilters ? AsyncCompletionData.UpdateSelectionHint.NoChange : AsyncCompletionData.UpdateSelectionHint.SoftSelected;
+            }
+
+            return new AsyncCompletionData.FilteredCompletionModel(
+                ImmutableArray<AsyncCompletionData.CompletionItemWithHighlight>.Empty, selectedItemIndex: 0,
+                filters, selection, centerSelection: true, uniqueItem: default);
+        }
+
+        private IEnumerable<AsyncCompletionData.CompletionItemWithHighlight> GetHighlightedList(
+            IEnumerable<ExtendedFilterResult> filterResults,
+            string filterText,
+            bool highlightMatchingPortions)
+        {
+            var highlightedList = new List<AsyncCompletionData.CompletionItemWithHighlight>();
+            foreach (var item in filterResults)
+            {
+                var highlightedSpans = highlightMatchingPortions
+                                        ? _completionHelper.GetHighlightedSpans(item.VSCompletionItem.DisplayText, filterText, CultureInfo.CurrentCulture)
+                                        : ImmutableArray<TextSpan>.Empty;
+                highlightedList.Add(new AsyncCompletionData.CompletionItemWithHighlight(item.VSCompletionItem, highlightedSpans.Select(s => s.ToSpan()).ToImmutableArray()));
+            }
+
+            return highlightedList;
+        }
+
+        private static ImmutableArray<AsyncCompletionData.CompletionFilterWithState> GetUpdatedFilters(
+            List<ExtendedFilterResult> filteredList,
+            ImmutableArray<AsyncCompletionData.CompletionFilterWithState> filters)
+        {
+            // See which filters might be enabled based on the typed code
+            var textFilteredFilters = filteredList.SelectMany(n => n.VSCompletionItem.Filters).Distinct();
+
+            // When no items are available for a given filter, it becomes unavailable
+            return ImmutableArray.CreateRange(filters.Select(n => n.WithAvailability(textFilteredFilters.Contains(n.Filter))));
+        }
+
+        private void MakeMostRecentItem(string item)
+        {
+            var updated = false;
+
+            while (!updated)
+            {
+                var oldItems = _recentItems;
+                var newItems = oldItems.Remove(item);
+
+                if (newItems.Length == MaxMRUSize)
+                {
+                    // Remove the least recent item.
+                    newItems = newItems.RemoveAt(0);
+                }
+
+                newItems = newItems.Add(item);
+                updated = ImmutableInterlocked.InterlockedCompareExchange(ref _recentItems, newItems, oldItems) == oldItems;
```

Comment:
Wat?  The list only has 10 items in it.  And we only call into this when a user commits something right?  So why do we need any sort of complex Interlocked exchanging?  just make a local for the field, add to it, and then assign that back into the field.  If you have any threading concerns, just take a lock.  We don't need any sort of fancy lock-free algorithms here that are so complex and hard to review.

Reply:
Sure!

---
In reply to: [232510712](https://github.com/dotnet/roslyn/pull/29016#discussion_r232510712) [](ancestors = 232510712)